### DISCUSSION TITLE
Custom EditCommandTemplate

### DISCUSSION THREAD

**Discussion by angelru** [Megabit/Blazorise#2163] [Q&A] [answered]
I have created a tabbed modal for editing and show it like this:

```
            <EditCommandTemplate>
                <Button Color="Color.Primary" Clicked="@(()=> ShowModal(context.Item))"><Icon Name="IconName.Edit" /> 
                </Button>
            </EditCommandTemplate>
```

if I change some field it is also changed in the table row before saving, how can I prevent that?

**Comment by angelru** [comment, marked as answer]
I understand that when displaying a modal (which is not from the datagrid) and passing the context to my modal, modifying it also changes the row (same object)

**Reply by stsrki** [reply]
You're basically bypassing the internal Datagrid editing? The thing that you experience is the reason why we have `context.CellValue` in `EditTemplate`. So that the editing value is stored in a temporary variable(s) and applied to the model only on save.

There is no other way to make it work unless you copy or clone the editing object or its values. And after you click save then re-apply them to the original object.

**Reply by angelru** [reply]
@stsrki this is what happens to me


https://user-images.githubusercontent.com/11645585/113771857-380e2500-9724-11eb-8fab-5517dd2568f0.mp4

**Reply by stsrki** [reply]
Yes, I'm aware. That's why I mentioned you need to clone the editing object and then use that for edit fields on the modal dialog.

**Reply by angelru** [reply]
and then how do I change the datagrid item? I can't use SaveCommand, resetting object from cloned context?

**Reply by stsrki** [reply]
Can you please post your entire page code so I can reproduce it and see what can be improved? It is hard to know what needs to be done like this.

**Reply by angelru** [reply]
If you need anything else tell me


Grid:

```
DataGrid TItem="CustomerResponse" Data="@customerResponses"
          ShowPager="true" PagerPosition="DataGridPagerPosition.Bottom"
          Editable="true">
    <DataGridColumns>
        <DataGridColumn TItem="CustomerResponse" Field="Persona.Nombre" Caption="@localizer["Name"]" />
        <DataGridColumn TItem="CustomerResponse" Field="Persona.Apellidos" Caption="@localizer["Surname"]" />
        <DataGridDateColumn TItem="CustomerResponse" Field="Persona.FechaNacimiento" DisplayFormat="{0:dd/MM/yyyy}" Caption="@localizer["Birthdate"]" />
        <DataGridColumn TItem="CustomerResponse" Field="Persona.NumDocumento" Caption="@localizer["DNI"]" />
        <DataGridColumn TItem="CustomerResponse" Field="Persona.TelefonoMovil" Caption="@localizer["Phone"]" />
        <DataGridColumn TItem="CustomerResponse" Field="Persona.EmailContacto" Caption="@localizer["Mail"]" />
        <DataGridColumn TItem="CustomerResponse" Field="Categoria.NombreCategoria" Caption="@localizer["Category"]" />
        <DataGridCommandColumn TItem="CustomerResponse" EditCommandAllowed="true" DeleteCommandAllowed="true" Caption="Actions">
            <NewCommandTemplate>
                @localizer["Actions"]
            </NewCommandTemplate>
            <EditCommandTemplate>
                <Button Color="Color.Primary" Clicked="@(()=> ShowModal(context.Item))"><Icon Name="IconName.Edit" /></Button>
            </EditCommandTemplate>
        </DataGridCommandColumn>
    </DataGridColumns>
</DataGrid>
```

Modal:
```
<Modal @ref="modalRef">
    <ModalContent Centered="true">
        <ModalHeader>
            <ModalTitle>@localizer["EditCustomer"] @selectedCustomerResponse.Persona.Nombre</ModalTitle>
            <CloseButton Clicked="@HideModal" />
        </ModalHeader>
        <ModalBody>
            <Tabs SelectedTab="@selectedTab" SelectedTabChanged="@OnSelectedTabChanged" 
                Margin="Margin.Is3.FromBottom">
                <Items>
                  ...........
                </Items>
                <Content>
                   ............
                </Content>
            </Tabs>
        </ModalBody>
        <ModalFooter>
            <Button Color="Color.Success" Clicked="@EditCustomerAsync"><Icon Name="IconName.Check" /></Button>
            <Button Color="Color.Secondary" Clicked="@HideModal"><Icon Name="IconName.Times" /></Button>
        </ModalFooter>
    </ModalContent>
</Modal>
```

```
  public async Task EditCustomerAsync()
    {
     //Call API
    }
```

**Reply by stsrki** [reply]
What is inside of ShowModal?

**Reply by angelru** [reply]
```
    private void ShowModal(CustomerResponse customerResponse)
    {
        selectedTab = "personal";
        selectedCustomerResponse = customerResponse

        modalRef.Show();
    }
```

**Reply by stsrki** [reply]
And then I would guess you use `selectedCustomerResponse` for TextEdit and other fields on Modal?

**Reply by stsrki** [reply]
Instead of assigning it can you create a new instance?

```cs
selectedCustomerResponse  = new CustomerResponse()
{
  Field1 = customerResponse.Field1,
  Field2 = customerResponse.Field2,
  Field3 = customerResponse.Field3,
  // etc.  
}
```

**Reply by angelru** [reply]
> Instead of assigning it can you create a new instance?
> 
> ```cs
> selectedCustomerResponse  = new CustomerResponse()
> {
>   Field1 = customerResponse.Field1,
>   Field2 = customerResponse.Field2,
>   Field3 = customerResponse.Field3,
>   // etc.  
> }
> ```

yes, I use selectedCustomerResponse although with automapper, because if not when changing something in the modal it also changes it in the grid automatically.

in this way, it is not changed in the grid row. But I still need that when I hit the save button it changes in the grid. I think I should get the row using linq and change the properties manually or replace the list object, is that the way to go?

**Reply by stsrki** [reply]
Or add both `selectedCustomerResponse ` and `editingCustomerResponse ` so that you can later save changes to `selectedCustomerResponse`.

```cs
private void ShowModal(CustomerResponse customerResponse)
{
  selectedTab = "personal";
  selectedCustomerResponse = customerResponse;

  // use editingCustomerResponse on the modal
  editingCustomerResponse  = new CustomerResponse()
  {
    Field1 = customerResponse.Field1,
    Field2 = customerResponse.Field2,
    // etc.  
  }

  modalRef.Show();
}
```

```cs
private void OnSave()
{

  // re-apply to the reference used by the grid
  selectedCustomerResponse.Field1 = editingCustomerResponse.Field1;
  selectedCustomerResponse.Field2 = editingCustomerResponse.Field2;
  // etc
}
```

**Reply by angelru** [reply]
> Or add both `selectedCustomerResponse ` and `editingCustomerResponse ` so that you can later save changes to `selectedCustomerResponse`.
> 
> ```cs
> private void ShowModal(CustomerResponse customerResponse)
> {
>   selectedTab = "personal";
>   selectedCustomerResponse = customerResponse;
> 
>   // use editingCustomerResponse on the modal
>   editingCustomerResponse  = new CustomerResponse()
>   {
>     Field1 = customerResponse.Field1,
>     Field2 = customerResponse.Field2,
>     // etc.  
>   }
> 
>   modalRef.Show();
> }
> ```
> 
> ```cs
> private void OnSave()
> {
> 
>   // re-apply to the reference used by the grid
>   selectedCustomerResponse.Field1 = editingCustomerResponse.Field1;
>   selectedCustomerResponse.Field2 = editingCustomerResponse.Field2;
>   // etc
> }
> ```

yes, that's the way to go, instead of mapping the properties one by one I used automapper, also ObservableRangeCollection so I can handle the collection with more options than List.

https://github.com/jamesmontemagno/mvvm-helpers/blob/master/MvvmHelpers/ObservableRangeCollection.cs

**Reply by stsrki** [reply]
Yeah, Automapper is the way to go. Here I just gave you a logic-way.

Good we finally solved it :)

**Comment by am185760** [comment]
https://stackoverflow.com/questions/75281786/multiple-functions-to-be-render-in-clicked-editcommandtemplate-datagrid-blazoris
Kindly help. I need a custom function for onlclick with context.clicked

**Reply by stsrki** [reply]
Do this

```razor
<EditCommandTemplate>
    <Button Color="Color.Primary" Clicked="@(()=>OnCustomEditCommand(context))">@context.LocalizationString</Button>
</EditCommandTemplate>
```

and

```razor
@code{
    async Task OnCustomEditCommand(EditCommandContext<Employee> context)
    {
        // do something here

        await context.Clicked.InvokeAsync();
    }
}
```

**Reply by am185760** [reply]
![image](https://user-images.githubusercontent.com/122779717/215496253-8c0d68c7-aa78-46e1-a17e-f5b2aed89caf.png)

How to set Edit / Delete btn position at the last coloumn

**Reply by stsrki** [reply]
Can you explain? How do you want to have them positioned?

**Reply by am185760** [reply]
as the last column..

**Reply by stsrki** [reply]
Ah. Just place the DataGridCommandColumn as the last tag as part of the DataGrid.

**Reply by am185760** [reply]
![image](https://user-images.githubusercontent.com/122779717/215564928-3727de3c-af27-4d7b-a9b5-0c1a53ba9f14.png)
I have done EditMode="DataGridEditMode.Popup".. is there any option to make them 3 coloumns in 1 row ?
like right now all columns are appearing on different rows..

**Reply by am185760** [reply]
I have used PopupFieldColumnSize="ColumnSize.Is3" to make field size small... but still all are appearing on different rows...

**Reply by stsrki** [reply]
Can you post the full code of your DataGrid usage?

**Reply by am185760** [reply]
@page "/NoteSetType"
@using EView360.Services;
@inject NoteSetTypeService service

<h3>NoteSetType</h3>


<DataGrid @ref="@dataGridRef"
          TItem="DbModels.NoteSetType"
          Data="@noteSetTypes"
          @bind-SelectedRow="@selectedNoteSetType"
          Editable
          EditMode="DataGridEditMode.Popup"
          Responsive="true"
          ShowPager
          RowInserted="@OnRowInserted"
          RowUpdated="@OnRowUpdated"
          RowRemoving="@OnRowRemoving"
          RowRemoved="@OnRowRemoved"
          UseValidation
          ShowValidationsSummary="false"
          ShowValidationFeedback
          PopupSize = "ModalSize.Large"
          Filterable
          Bordered>

    <DataGridCommandColumn CellClass="@((item)=> "all_headers")">
        <NewCommandTemplate>
            <Button Color="Color.Success" Clicked="@context.Clicked">New</Button>
        </NewCommandTemplate>
        <EditCommandTemplate>
            <Button Color="Color.Primary" Clicked="@(()=>OnEditBtn(context.Item))">Edit</Button>
        </EditCommandTemplate>
        <DeleteCommandTemplate>
            <Button Color="Color.Danger" Clicked="@context.Clicked">Delete</Button>
        </DeleteCommandTemplate>
        <CancelCommandTemplate>
            <Button Color="Color.Secondary" Clicked="@(()=>OnCancelBtn(context.Item))">Cancel</Button>
        </CancelCommandTemplate>
        <SaveCommandTemplate>
            <Button Color="Color.Success" Clicked="@context.Clicked">Save</Button>
        </SaveCommandTemplate>
    </DataGridCommandColumn>

    <DataGridSelectColumn Displayable="false" Field="@nameof(DbModels.NoteSetType.CreatedByName)" CustomFilter="@CustomUserFilter" Caption="Created By" Editable>
        <FilterTemplate>
            <Select Class="form-select" TValue="string" SelectedValue="@selectedUser" SelectedValueChanged="@(value => { selectedUser = value; context.TriggerFilterChange(selectedUser); })">
                <SelectItem Value="@("*")">Any</SelectItem>
                @foreach (string userName in noteSetTypes.Select(x => x.CreatedByName).ToList().Distinct())
                {
                    <SelectItem Value="@(userName)">@userName</SelectItem>
                }
            </Select>
        </FilterTemplate>
    </DataGridSelectColumn>

    

    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is12" Field="@nameof(DbModels.NoteSetType.NoteSetTypeName)" Validator="@CheckName" Caption="Name" Sortable="false" Editable CellClass="@((item)=> "otherFields")" />
     
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType1Title)" Caption="Type1 Title" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridNumericColumn PopupFieldColumnSize="ColumnSize.Is3" Field="@nameof(DbModels.NoteSetType.DenominationType1)" Caption="Type1" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType1MultiCurrency)" Caption="IsType1MultiCurrency" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType1Recycler)" Caption="IsType1Recycler" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />


    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType2Title)" Caption="Type2 Title" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Field="@nameof(DbModels.NoteSetType.DenominationType2)" Caption="Type2" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType2MultiCurrency)" Caption="IsType2MultiCurrency" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType2Recycler)" Caption="IsType2Recycler" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />

    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType3Title)" Caption="Type3 Title" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Field="@nameof(DbModels.NoteSetType.DenominationType3)" Caption="Type3" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType3MultiCurrency)" Caption="IsType3MultiCurrency" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />
    <DataGridColumn PopupFieldColumnSize="ColumnSize.Is3" Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType3Recycler)" Caption="IsType3Recycler" Editable Filterable="false" CellClass="@((item)=> "otherFields")" />

  @*  <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType4Title)" Caption="Type4 Title" Editable Filterable="false" />
    <DataGridColumn Field="@nameof(DbModels.NoteSetType.DenominationType4)" Caption="Type4" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType4MultiCurrency)" Caption="IsType4MultiCurrency" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType4Recycler)" Caption="IsType4Recycler" Editable Filterable="false" />

    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType5Title)" Caption="Type5 Title" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType5)" Caption="Type5" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType5MultiCurrency)" Caption="IsType5MultiCurrency" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType5Recycler)" Caption="IsType5Recycler" Editable Filterable="false" />

    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType6Title)" Caption="Type6 Title" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType6)" Caption="Type6" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType6MultiCurrency)" Caption="IsType6MultiCurrency" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType6Recycler)" Caption="IsType6Recycler" Editable Filterable="false" />

    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType7Title)" Caption="Type7 Title" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.DenominationType7)" Caption="Type7" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType7MultiCurrency)" Caption="IsType7MultiCurrency" Editable Filterable="false" />
    <DataGridColumn Displayable="@isEditMode" Field="@nameof(DbModels.NoteSetType.IsType7Recycler)" Caption="IsType7Recycler" Editable Filterable="false" />*@

</DataGrid>




<style>
    form {
        max-width: 1200px;
    }

    .all_headers {
        width: 170px;
        overflow: hidden;
        /*display: inline-block;*/
        white-space: nowrap;
    }

    .btn-danger {
        margin-left: 5px;
    }
</style>

@code {
    private List<DbModels.NoteSetType> noteSetTypes = new();
    private DbModels.NoteSetType? selectedNoteSetType;
    private bool isEditMode = false;
    private DataGrid<DbModels.NoteSetType>? dataGridRef;
    string? selectedUser;

    protected override async void OnInitialized()
    {
        noteSetTypes = await service.GetNoteSetTypeListAsync();
        await base.OnInitializedAsync();
    }

    private string HeaderCellClass()
    {
        return "tableHeader";
    }

    private bool CustomUserFilter(object itemValue, object searchValue)
    {
        if (searchValue is string userFilter)
        {
            return userFilter == "*" || userFilter == itemValue?.ToString();
        }

        return true;
    }

    public void CheckName(ValidatorEventArgs validationArgs)
    {
        ValidationRule.IsNotEmpty(validationArgs);
        if (validationArgs.Status == ValidationStatus.Error)
        {
            validationArgs.ErrorText = "This field is required";
        }
        else if (noteSetTypes.Count(x => x.NoteSetTypeName.Trim().ToLower() == validationArgs.Value.ToString().Trim().ToLower()) > 1)
        {
            validationArgs.ErrorText = "Note type name already exists. Please provide unique name";
            validationArgs.Status = ValidationStatus.Error;
        }
    }
    private async Task OnEditBtn(DbModels.NoteSetType noteSetType)
    {
        isEditMode = true;
        await dataGridRef.Edit(noteSetType);
    }
    private async Task OnCancelBtn(DbModels.NoteSetType noteSetType)
    {
        isEditMode = false;
        await dataGridRef.Cancel();
    }

    async Task OnRowInserted(SavedRowItem<DbModels.NoteSetType, Dictionary<string, object>> savedRow)
    {
        savedRow.Item.CreationTime = DateTime.Now;
        service.InsertNoteSetType(savedRow.Item);
    }
    async Task OnRowUpdated(SavedRowItem<DbModels.NoteSetType, Dictionary<string, object>> savedRow)
    {
        isEditMode = false;
        service.UpdateNoteSetType(savedRow.Item, savedRow.Item.NoteSetTypeId);
    }
    async Task OnRowRemoving(CancellableRowChange<DbModels.NoteSetType> cancellableRow)
    {
        if (service.IfAtmExistForNoteSetType(cancellableRow.Item.NoteSetTypeId).Result)
        {
            cancellableRow.Cancel = true;
        }
    }
    async Task OnRowRemoved(DbModels.NoteSetType noteSetType)
    {
        service.DeleteNoteSetType(noteSetType.NoteSetTypeId);
    }
}

**Reply by am185760** [reply]
[noteSetType component.txt](https://github.com/Megabit/Blazorise/files/10539435/noteSetType.component.txt)

kindly check this code

**Reply by am185760** [reply]
<DataGridColumn Field="@nameof( Employee.Email )" Caption="Email" Editable EditFieldColumnSize="ColumnSize.IsFull.OnDesktop" />
I followed your example

**Reply by am185760** [reply]
Hi @stsrki  ... any luck related to modal field adjustment ?
otherwise I will be creating a separate bootstrap modal writing its all code from scratch..

**Comment by am185760** [comment]
@stsrki  issue resolved .. 
I replaced  `<link rel="stylesheet" href="css/bootstrap/bootstrap.min.css" />`

By

` <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.1/dist/css/bootstrap.min.css" integrity="sha384-zCbKRCUGaJDkqS1kPbPd7TveP5iyJE0EjAuZQTgFLD2ylzuqKfdKlfG/eSrtxUkn" crossorigin="anonymous">`

**Reply by stsrki** [reply]
Sorry for not responding sooner. I went to bed early last night.

Thanks for posting the solution.

**Comment by am185760** [comment]
@stsrki  https://stackoverflow.com/questions/75461612/unable-to-set-timepicker-in-blazorise-blazor  please help in this one.. thanks

**Reply by stsrki** [reply]
You're trying to initialize TimePicker with a `new` keyword and use it like an ordinal class, Which is not going to work. Your need to define it declaratively in a `.razor` file. Or, if you want to use it dynamically, then you need to use it in a RenderFragment and render it manually https://learn.microsoft.com/en-us/aspnet/core/blazor/advanced-scenarios?view=aspnetcore-7.0
